Export and merge Pontoon terminology for Smartling

bcolsson opened the pull request.

bcolsson committed:
Added working tbx file merging script

bcolsson committed:
Delete .DS_Store

bcolsson committed:
Compare pontoon export with smartling, update IDs to match for import

bcolsson committed:
Finished logic for importing into Smartling. CLI arguments.

bcolsson committed:
Formatted with black

bcolsson committed:
Delete .DS_Store (#25)

bcolsson committed:
Added Pontoon tbx file export and modified cli argruments.

bcolsson committed:
Formatted with black, changed script name

flodolo commented on Pontoon/tbx_merge/tbx_merge.py:
```diff
@@ -0,0 +1,176 @@
+        ).text
+        try:
+            definition = termEntry.find(
+                "./langSet[@{http://www.w3.org/XML/1998/namespace}lang='en-US']/descripGrp/descrip[@type='definition']"
```
You can also define a namespace once, and make this a bit more readable

```
nsmap = {"xml": "http://www.w3.org/XML/1998/namespace"}
term = termEntry.find("./langSet[@xml:lang='en-US']/ntig/termGrp/term", nsmap).text
```

flodolo commented on Pontoon/tbx_merge/tbx_merge.py:
```diff
@@ -0,0 +1,176 @@
+            ).text
+        except AttributeError:
+            definition = None
+        pontoon_term_map[(term, definition)] = termEntry
```
Interesting, didn't realize a TBX can have multiple terms with a different definition, and I don't think I've ever used tuples as dictionary keys.

mathjazz replied:
Yeah, and we actually make use of that TBX feature:
https://pontoon.mozilla.org/it/terminology/common/?list=212229,212230,212243,212244,212245,212246,212292,212293

flodolo commented on Pontoon/tbx_merge/tbx_merge.py:
```diff
@@ -0,0 +1,176 @@
+def replace_pontoon_ids(etree, smartling_dict):
+    """Replaces Pontoon IDs with Smartling IDs if the term and definition match exactly a term in Smartling glossary file."""
+    root = etree.getroot()
+    pontoon_term_map = {}
```
nit: it's a `map` here, but it's a `dict` in `smartling_dict`, should probably be the same.

flodolo commented on Pontoon/tbx_merge/locales.txt:
```diff
@@ -0,0 +1,5 @@
+ja
```
Alphabetical order, please ;-)

mathjazz commented:
Looks good!

Just a quick thought on API and the arguments: If the general use case will be to only generate one export (with either existing, new or smartling IDs), you could replace the `--new` and `--merge-only` arguments with one called something like `--id-strategy`. The possible values would be `keep`, `remove` and `smartling`, the latter one requiring the `--smartling` argument.

bcolsson commented:
Thanks for the feedback! I've implemented it all here. Didn't seem to break anything so also moved this branch to the pontoon-scripts repo.
https://github.com/bcolsson/scripts/pull/27
